### PR TITLE
google oauth setup

### LINKED ISSUE #204: Google Login - Backend
- access google console with a blueprint email

- implement logic to detect wether user entered email is a .theshoeproject.online domain
    - .uwblueprint.org email is also valid

- create the user in the backend if they are valid
  - if the email is invalid redirect to an error page
  - if its valid redriect to all stories page (random page for now)
      - generate a JWT token & store _somewhere_ where we can validate it (session cookie?) and return to frontend

### PR DESCRIPTION
**Pull request by s278gupt**
Closes: https://github.com/uwblueprint/shoe-project/issues/204

- Set up project on google api console
- Create `/login` endpoint
  - Sets `state` and redirects to google oauth endpoint
  - Google oauth set up to call our callback endpoint below
- Create `/auth/callback` endpoint
  - Checks state 
  - Checks email domain
  - Retrieves user email
  - Encodes + returns jwt token (setting jwt token in local session storage is for next pr)
- If unauthorised, redirects to vanilla `/unauthorized` endpoint
- Updates some of the frontend boilerplate setup (due to changes in `User` model)

<b>To reviewers:</b>
Feel free to tophat by going to `localhost:8900/api/login` (if you're unable to login to Google you might need to be added to our project on the google console -- please slack me!)

### CODE REVIEW

File: README.md
Comment (AbhiPrasad) [inline README.md:118]: Let's make sure that we have a TODO somewhere here to add back the Auth docs

File: restapi/auth.go
Comment (AbhiPrasad) [inline restapi/auth.go]: We can't declare this outside the function (maybe we should add it as a field to the api struct)

File: restapi/auth.go
Comment (AbhiPrasad) [inline restapi/auth.go]: Should we make this a struct method on user?

```go
func (user u) hasValidEmail() bool {
    return user.Hd == "uwblueprint.org" || user.Hd == "theshoeproject.online"
}
```

that way if we decide to change `Hd` in the future, we don't have to reach into the API

File: restapi/endpoints_test.go
Comment (AbhiPrasad) [inline restapi/endpoints_test.go:306]: Wait the endpoint doesn't need a header anymore?

File: restapi/auth.go
Comment (AbhiPrasad) [inline restapi/auth.go]: > in prod, set Secure to true

Can we do this by looking at the ENV variables for the time being?

File: restapi/auth.go
Comment (AbhiPrasad) [inline restapi/auth.go]: can we `rm -rf` the println?

File: internal/database/models/auth.go
Comment (AbhiPrasad) [inline internal/database/models/auth.go:15]: Hmm I'm wondering if we can write a CHECK constraint here to limit the Hd values to one of the domains, but idk how hard it is with gorm.

File: restapi/endpoints_test.go
Comment (s278gupt) [reply restapi/endpoints_test.go:306]: No, I moved the endpoints around so this is no longer protected (temporarily) and the tests are updated accordingly so they don't fail (the tests will be all updated once at the end of this auth epic)

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go]: good call

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go]: i actually think it's the opposite...if things change in the future, we don't want to reach into database-layer logic? @nimwijetunga what do you think?

File: restapi/auth.go
Comment (nimwijetunga) [inline restapi/auth.go:49]: Just so I understand the `r.FormValue("state")` is added on by google right? We don't have to include anything on our frontend?

File: restapi/auth.go
Comment (nimwijetunga) [reply restapi/auth.go]: Well I actually think @AbhiPrasad suggestion makes sense. Since this logic is custom to us (i.e has nothing to do with the actual google OAUTH flow) it would be better to make it a member of the User model to encapsulate that logic.

File: restapi/auth.go
Comment (nimwijetunga) [reply restapi/auth.go:49]: Same for `r.FormValue("code")` below right?

File: restapi/auth.go
Comment (nimwijetunga) [inline restapi/auth.go:84]: I thought we were going to put the JWT token in the cookie of the response?

Comment (nimwijetunga) [reply restapi/auth.go:84]: Also we should redirect users to the all stories page if the authentication is successful. Since this page currently doesn't exist we can redirect them to `/` (i.e homepage) and create another (separate) pr to change this once all stories has been implemented.

File: restapi/auth.go
Comment (nimwijetunga) [inline restapi/auth.go:75]: This should be `/unauthorized` (i.e remove the api prefix). This page should exist when [this pr](https://github.com/uwblueprint/shoe-project/pull/223) is merged in.

File: restapi/redirects.go
Comment (nimwijetunga) [inline restapi/redirects.go:1]: We don't need this since we can redirect directly to `/unauthorized` which is a frontend page

**Review by nimwijetunga — CHANGES_REQUESTED** [review]
Just a few small things to fix, looks good overall!

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go:84]: ok, this will be done in next pr! didn't want to touch jwt flow in the same pr...

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go:75]: will be changed in next pr (for jwt flow) so I don't have to wait for Dinu's pr to be merged.

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go:49]: yep, this callback endpoint is only ever called by google; all the query params are set by them too

File: internal/database/models/auth.go
Comment (s278gupt) [reply internal/database/models/auth.go:15]: see above

File: restapi/auth.go
Comment (s278gupt) [reply restapi/auth.go]: Ok, I've added this logic in a hook before any user record is created (see latest commit)

File: config/getter.go
Comment (nimwijetunga) [inline config/getter.go:178]: Can you add these to heroku as well otherwise we might forget

File: restapi/auth.go
Comment (nimwijetunga) [inline restapi/auth.go:62]: Maybe we should extract this URL into a constant?

**Review by nimwijetunga — APPROVED** [review]
LGTM! Please make sure you test this to see if the flow works before merging!

**Review by AbhiPrasad — APPROVED** [review]
Make sure we run `go mod tidy` and `make backend-fmt` before merge

File: config/getter.go
Comment (s278gupt) [reply config/getter.go:178]: https://app.zenhub.com/workspaces/the-shoe-project-f20-5f729b8f69fd6b2517b9a848/issues/uwblueprint/shoe-project/243

**Pull request merged**
